File: 0124. Binary Tree Maximum Path Sum.java
// Hard
// DFS, Greedy
// Time:O(n), Space:O(h)
// https://leetcode.cn/problems/binary-tree-maximum-path-sum/

class Solution {
    // Use DFS to calculate max Pathsum for each node from its left/right
    // Greedy: ignore negative contributions, maintain the maxSum to record them
    
    int maxSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        maxGain(root);
        return maxSum;
    }
    public int maxGain(TreeNode node) {
        if (node == null) return 0;
        int leftGain = Math.max(maxGain(node.left), 0);
        int rightGain = Math.max(maxGain(node.right), 0);
        int curMax = node.val + leftGain + rightGain;
        maxSum = Math.max(maxSum, curMax);
        // Return max Gain to parent node
        return node.val + Math.max(leftGain, rightGain);
    }
}

File: 0124. Binary Tree Maximum Path Sum.py
from typing import *
from TreeNode import TreeNode

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_sum = float('-inf')
        self.maxGain(root)
        return self.max_sum
    
    def maxGain(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0
        
        left = max(self.maxGain(node.left), 0)
        right = max(self.maxGain(node.right), 0)
        
        self.max_sum = max(self.max_sum, node.val + left + right)

        return node.val + max(left, right)
